### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `DecoupleTypeCast` silently clobbers vector lanes the source left unwritten when a cast-store sits under a nested-`if` (or `if/else`) guard

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2616]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11 (source at `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`).

### System information

Reproduced on NVIDIA H100 80GB HBM3 (sm_90), CUDA 13, tilelang 0.1.11 (dev wheel, byte-identical to the source tag above). `DecoupleTypeCast` is a target-independent TIR transform (runs in the CUDA/ROCm/CPU/Metal pipelines), so the defect is not GPU-specific; reproduced on sm_90 this session.

### Problem description

`DecoupleTypeCast` splits a vectorized mixed-precision cast-store loop into `copy-from → compute → copy-to` loops, but recovers only a **single top-level `if`-without-`else`** guard for the copy loops — so when the store sits under any other guard shape (a nested `if`, or an `if/else`), the copy-to loop it emits is **weaker** than the store it replaces and silently overwrites buffer lanes the source deliberately left unwritten with uninitialized register data. The kernel compiles and runs with no error or warning; the result is a silent wrong write to valid, in-bounds buffer indices (so `compute-sanitizer` is blind to it).

Concretely, for a `T.vectorized(16)` cast-store guarded by `if i < 12: if i < 8: B[i] = cast(...)`, the transformed copy-to loop is guarded by `i < 12` only — the inner `i < 8` is dropped — so lanes 8–11 (which the source never wrote) are clobbered.

<details><summary>Trigger boundary — the two guard shapes that lose the guard, vs the shape that's handled (source-level; only the nested case was run this session)</summary>

`extract_if_condition` recovers a guard only for a single top-level `IfThenElse` whose `else_case is None`. The other shapes fall through to `condition = None`, and the copy-to loop is emitted with a weaker guard than the store:

| store guard shape | recovered copy-guard | copy-to loop vs store |
|---|---|---|
| single `if`, no `else` (the handled case) | the `if` condition | equal — correct |
| **nested `if`** (`if a: if b: store`) — TESTED | outer condition only (`a`) | weaker — inner `b` dropped, lanes satisfying `a & !b` clobbered |
| **`if/else`** (store in one branch) — inferred from source, not run this session | `None` | unconditional — all lanes clobbered |

Only the nested-`if` face was executed this session (below). The `if/else` face is consistent with the same `extract_if_condition` bail (`else_case is None` fails → `None`) but is stated from source reading, not a run.

</details>

### Reproducible example code

```python
import torch, tilelang
import tilelang.language as T
N = 16

@tilelang.jit
def kf():
    @T.prim_func
    def main(A: T.Tensor((N,), T.float8_e4m3fn), B: T.Tensor((N,), T.float8_e4m3fn)):
        with T.Kernel(1, threads=32):
            af = T.alloc_local((N,), T.float32)
            T.copy(A, af)
            for i in T.vectorized(N):
                if i < 12:
                    if i < 8:
                        B[i] = T.cast(af[i], T.float8_e4m3fn)   # source writes only i<8
    return main

# CONTROL: identical kernel but B is float32 (same dtype as af) -> no Cast -> DecoupleTypeCast does not fire
@tilelang.jit
def kf_ctrl():
    @T.prim_func
    def main(A: T.Tensor((N,), T.float32), B: T.Tensor((N,), T.float32)):
        with T.Kernel(1, threads=32):
            af = T.alloc_local((N,), T.float32)
            T.copy(A, af)
            for i in T.vectorized(N):
                if i < 12:
                    if i < 8:
                        B[i] = af[i]
    return main

a = (torch.arange(N, device="cuda", dtype=torch.float32) * 0.5).to(torch.float8_e4m3fn)
b = torch.full((N,), 7.0, device="cuda").to(torch.float8_e4m3fn)   # sentinel, exact in fp8
kf()(a, b)
print("bug  B =", b.to(torch.float32).tolist())
# -> [0,0.5,1,1.5,2,2.5,3,3.5, 0,0,0,0, 7,7,7,7]
#    B[8:12] CLOBBERED to 0.0 (should stay 7.0 — source never wrote them); B[12:16]=7.0 (outside i<12)

a2 = torch.arange(N, device="cuda", dtype=torch.float32)
b2 = torch.full((N,), 7.0, device="cuda", dtype=torch.float32)
kf_ctrl()(a2, b2)
print("ctrl B =", b2.tolist())
# -> [0,1,2,3,4,5,6,7, 7,7,7,7, 7,7,7,7]   B[8:16] all preserved (pass did not fire)
```

Observed this session (H100 sm_90): bug `B = [0,0.5,1,1.5,2,2.5,3,3.5, 0,0,0,0, 7,7,7,7]`; control `B[8:16]` all `7.0`. The only difference between the two runs is the presence of the dtype cast that triggers `DecoupleTypeCast`, isolating it as the culprit.

### Traceback

```pytb
No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic. (All writes are in-bounds, so `compute-sanitizer` reports nothing.)
```

### Expected behavior

The transform should be semantics-preserving: a lane the original store did not write must remain unwritten. The single-`if`-without-`else` path is already correct (the copy loops are guarded by the recovered condition); the nested-`if` and `if/else` shapes should reach the same faithful guarding — or, if the guard cannot be represented, the transform should conservatively skip (leave the loop untransformed, which is always correct, as the same-dtype and single-`if` paths show the pass is optional).

### Additional context

**Root cause.** `DecoupleTypeCast` mis-recovers the store's guard for any `if`-shape other than a single top-level `if`-without-`else`, so the copy-to loop it emits is weaker than the store it replaces and overwrites lanes the source left unwritten. Mechanism: [`extract_if_condition` (`decouple_type_cast.py:278-286`)](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/tilelang/transform/decouple_type_cast.py#L278-L286) returns `(stmt.condition, then)` **only** when `isinstance(stmt, IfThenElse) and stmt.else_case is None` (line 284), else `(None, stmt)`. For a nested `if` it peels only the outer condition (returning the inner `if` as the body, whose guard is then lost); for an `if/else` it returns `None`. `visit_for_` (line 398) calls it once; the **compute** loop is built from the whole `normalized_body` (lines 428-431) so it keeps *both* guards, but [`_create_copy_loops` (lines 488-539)](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/tilelang/transform/decouple_type_cast.py#L488-L539) wraps the copy store in a guard only when `condition is not None` (line 523) — so with `condition = None` (if/else) the copy-to loop runs for the full extent, and with `condition = outer-only` (nested) it runs under just the outer guard. In both cases `b_local_cast[lane]` is written by the compute loop only for the fully-guarded lanes, but the copy-to loop writes `b[lane] = b_local_cast[lane]` for the wider set, so the extra lanes copy uninitialized register values into the buffer.

<details><summary>Emitted CUDA for the nested-<code>if</code> repro (confirms the dropped inner guard)</summary>

Compute loop keeps both guards (writes lanes 0–7 only); copy-to loop keeps only the outer `i < 12`:

```cuda
// compute:  for (i_1 < 4) if (i_1 < 3) if (i_1 < 2) { B_local_cast_1[...] = cast(...) }   // lanes 0-7
// copy-to:  for (i_copy_s < 16) if (i_copy_s < 12) { B[i_copy_s] = B_local_cast_1[i_copy_s] }  // inner <8 DROPPED
```

Lanes 8–11 satisfy `i_copy_s < 12` but `B_local_cast_1[8:11]` was never written (reads 0.0), so `B[8:12]` is clobbered; lanes 12–15 fall outside `i < 12` and keep the 7.0 sentinel.

</details>

**Suggested fix.** `extract_if_condition` should not silently discard a guard it cannot represent. Minimal-safe option: in `visit_for_`, after `condition, _ = extract_if_condition(...)`, if the normalized body still contains an `IfThenElse` guarding the store that the recovered `condition` does not fully capture, **skip the transform** (leave the loop untransformed — always correct, per the same-dtype and single-`if` paths). Fuller option: recognize nested `IfThenElse` chains and conjoin their conditions (and, for `if/else`, restrict to the branch containing the store).

**Provenance.** The `copy-from/compute/copy-to` split with this single-top-level-`if` guard recovery is how `DecoupleTypeCast` has worked since it shipped; the missing nested/`else` handling is a coverage gap in `extract_if_condition`, not a regression of previously-correct behavior. (Live on `main` — `extract_if_condition` unchanged.)

**Dedup.** I searched the open and closed tracker for `DecoupleTypeCast` / decouple type cast and found no existing report. Distinct from the atomic-read-reorder issues (#2594 `IfStmtBinding`, #2595 pipeline) — same "a transform emits code weaker than the source under a guard" family, but a different pass and a different mechanism (guard recovery in a vectorized cast-store split, not a bind/read reorder).
